Make DataGateway Work with API's ICAT Backend

MRichards99 opened the pull request:
## Description
In a sentence, these changes allow DataGateway to use DataGateway API's Python ICAT backend, the backend to be used in production. The majority of these changes revolve around moving attribute names from SNAKE_CASE to camelCase, so many of these commits (especially a2c7fde001a4f10456aead012951db029737a495) are the result of find and replace. However, there foreign key attributes have been removed (e.g. `INVESTIGATION_ID` on `Dataset`) as the ICAT backend doesn't make use of foreign keys. 

There's one _very_ minor change in functionality - full names of users are now displayed as per b7f0642a3fa0492ca9f2948690e856351ed824c0.

I've left a couple of TODOs in (search for `TODO -` which will display mine) just to serve as reminders about a couple things. They should probably get removed before getting merged, but I wanted to keep them there until _everything_ passes etc.

On that note, there are still a few outstanding things on this branch:
- My data table views (both ISIS and Diamond) have issues on their e2e tests. There is an issue for this (#659)
- Lots of e2e tests on Search fail due to a slow query involving facilityCycles that are included in said query. Again, there is an issue for this (#658) (EDIT: this is fixed as of 26/05/21)
- On dataview, e2e tests on Diamond proposals (card & table view) fail - this is because ICAT Server 4.11.1 needs to be installed on preprod (an request to the API with a distinct filter on multiple fields which needs the newer version of ICAT Server)
- `cartSelection.spec.ts` fails due to issues with the my data views

The CI will inevitably fail on the e2e tests because the DB backend is on preprod, rather than ICAT backend. I've made some screenshots to show the current state of the tests:

Unit Tests for all components- all pass nicely:
![all-unit-tests](https://user-images.githubusercontent.com/32678030/115035457-81006f00-9ec4-11eb-8dec-f4e85f824ce2.JPG)

DataView Card View e2e - all pass except for Diamond proposals (4.11.1) as above
![dataview-e2e-batch1](https://user-images.githubusercontent.com/32678030/115035821-de94bb80-9ec4-11eb-8826-82ec0b8f29c2.JPG)

DataView Table View e2e - all pass except for Diamond proposals (4.11.1) and mydata (issues above)
![dataview-e2e-batch2](https://user-images.githubusercontent.com/32678030/115035971-fc622080-9ec4-11eb-9296-0aa061926900.JPG)

DataView Remaining e2e - all pass except for cartSelection (mydata)
![dataview-e2e-batch3](https://user-images.githubusercontent.com/32678030/115036137-24ea1a80-9ec5-11eb-959c-ae18be7047b4.JPG)

Download e2e - all pass
![download-e2e](https://user-images.githubusercontent.com/32678030/115036340-56fb7c80-9ec5-11eb-8ac7-e37c43575f24.JPG)

Search e2e - lots of fails due to facilityCycles include (issue above)
![search-e2e](https://user-images.githubusercontent.com/32678030/115036434-6bd81000-9ec5-11eb-96b3-f282a3db4baa.JPG)

## Testing instructions
To test this, the ICAT backend of the API should be used. Specifically, use `temp-distinct-client-handling-merge` branch on DataGateway API - this has a mixture of the new client handling solution and the fix for the distinct filter and is the branch that's most like the API that'll be run in production at time of writing. For client handling configuration, I recommend the following settings for testing. You should be able to get away with `client_pool_init_size: 3` but the below settings are the ones that have actually been tested with this frontend branch:
```json
"client_cache_size": 10,
"client_pool_init_size": 7,
"client_pool_max_size": 10,
```

As mentioned above, ICAT Server 4.11.1 is required for the API from now on.

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage
- [ ] {more steps here}

## Agile board tracking
connect to #94

louise-davies commented on packages/datagateway-download/src/downloadApi.ts:
```diff
@@ -361,10 +361,11 @@ export const getDatafileCount: (
               eq: entityId,
             },
           },
+          include: 'dataset',
```
Adding this `include: dataset` causing this function to fail on Diamond prod - it's giving an error to do with ICAT not supporting multiple attributes? Regardless, the `where` on it's own works fine for the `python-icat` backend (although I presume not for the database backend)

Can we investigate this using DLS preprod to double check?

MRichards99 replied:
Does Diamond prod have ICAT Server 4.11.1? That might be why there's an issue

louise-davies replied:
It doesn't no - in fact I forgot that currently lots of requests currently fail due to this! So it's not a DataGateway issue then - perhaps we might need to use an older version of datagateway-api for DLS if the ICAT 4.11.1 release is delayed

MRichards99 replied:
I could certainly find the most recent API commit not to require 4.11.1 if needed, but that'll mean the distinct filter will be broken on that version of the API, something which Diamond need I think?

louise-davies replied:
This is more thinking of a worst case scenario in which we develop a work around in DataGateway to the issue (similar to what TopCAT does) then in that case we'd also need to drop the datagateway-api to a previous commit yeah. Nothing to worry about now though hopefully!

MRichards99 replied:
Yeah if we need to do something like that, just let me know and I'm sure we can get it sorted :)

MRichards99 commented:
I removed the failing test for the landing page work (and put studyInvestigation's investigation to mandatory/removed the `?`). I've also added the fixes for the datafiles stuff, and made the suggested refactor - a3fe16a7a30c826e67481fb2c97997ba3fc476c5 for this :)

MRichards99 commented:
CI shows the unit tests all pass, but I've run e2e tests locally to check them:

No changes for table view - my data still fail (existing issue for this), diamond proposals passed when I ran it in interactive mode, visits and the two failing investigations files were fixed in 9a8bc7dfcc055646375f2e5bab3a62145597a250.
![2405](https://user-images.githubusercontent.com/32678030/119380436-9c774a80-bcb8-11eb-976b-f6e878903ce6.JPG)

Card view are all good
![image](https://user-images.githubusercontent.com/32678030/119380630-e2341300-bcb8-11eb-9637-4cfaa29bc813.png)

Same with landing pages
![image](https://user-images.githubusercontent.com/32678030/119380645-e829f400-bcb8-11eb-9aab-cec6589b70c0.png)

Other dataview tests pass too (only cart selection ones to fail are the ISIS my data table)
![image](https://user-images.githubusercontent.com/32678030/119382173-9766cb00-bcb9-11eb-9268-7c441f555f93.png)

Download passes too
![image](https://user-images.githubusercontent.com/32678030/119383079-d6e1e700-bcba-11eb-9d78-88d4d2a55058.png)

Search still fails but hopefully it'll be fixed with the existing issue

sam-glendenning commented:
> Search still fails but hopefully it'll be fixed with the existing issue

In the middle of trying to get this fix done, @MRichards99. Currently having problems with merge conflicts between my branch and yours, causing a bit of a headache

sam-glendenning commented:
> > Search still fails but hopefully it'll be fixed with the existing issue
> 
> In the middle of trying to apply this fix, @MRichards99. Currently having problems with merge conflicts between my branch and yours, causing a bit of a headache

MRichards99 commented:
@sam-glendenning yeah I noticed the PR, great to hear it's close to being done! We can go through the merge conflicts together at some point if that helps :)

sam-glendenning commented:
> @sam-glendenning yeah I noticed the PR, great to hear it's close to being done! We can go through the merge conflicts together at some point if that helps :)

I'll keep trying to figure it out, if nothing else it's not a lot of code and I could just commit it directly to this branch without bothering with a PR for merging into this branch
